File: ME4/queue_linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node {
    int data;
    struct node *next;
} node;

typedef struct linked_list {
    struct node *head;
    struct node *tail;
} queue;

node* new_node(int data) {
    node *z;
    z = malloc(sizeof(struct node));
    z->data = data;
    z->next = NULL;

    return z;
}

queue* new_queue() {
    queue *q = malloc(sizeof(queue));
    q->head = NULL;
    q->tail = NULL;

    return q;
}

void traversal(queue *q) {
    node *temp = q->head; 

    while(temp != NULL) { 
        printf("%d ", temp->data);
        temp = temp->next;
    }

    printf("\n");
}

int is_empty(queue *q) {
    if(q->head == NULL)
        return 1;
    return 0;
}

void enqueue(queue *q, node *n) {
    if(is_empty(q)) {
        q->head = n;
        q->tail = n;
    }
    else {
        q->tail->next = n;
        q->tail = n;
    }
}

int dequeue(queue *q) {
    if(is_empty(q)) {
        traversal(q);
        return -1000;
    }
    else {
        int x = q->head->data;
        node *temp = q->head;
        q->head = q->head->next;
        free(temp);
        return x;
    }
}

int main() {
    queue *q = new_queue();
    int choice;

    while(1) {
        node* a;
        int input;
        printf("\nWhat do you want to do?\n1) Insert integer.\n2) Pop integer.\n3) Exit\n");
        scanf("%d", &choice);
        switch(choice) {
            case 1: 
                printf("Enter integer to insert: ");
                scanf("%d", &input);
                a = new_node(input);
                enqueue(q, a);
                printf("Inserted. Queue is now: \n");
                traversal(q);
                break;
            case 2:
                printf("Popping first in queue...\n");
                dequeue(q);
                printf("Popped. Queue is now:\n");
                traversal(q);
                break;
            default:
                exit(0);
                break;
        }
    }

    return 0;
}

File: ME4/queue_linked_list.py
class Node():
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue():
    def __init__(self):
        self.head = None
        self.tail = None

def traversal(q):
    temp = q.head 

    a = ''
    while temp != None: 
        a = a + str(temp.data) + ' '
        temp = temp.next

    print(a)  

def is_empty(q):
    if q.head == None:
        return True
    return False

def enqueue(q, n):
    if is_empty(q):
        q.head = n
        q.tail = n
    else:
        q.tail.next = n
        q.tail = n

def dequeue(q):
    if is_empty(q):
        traversal(q)
        return -1000
    else:
        x = q.head.data
        temp = q.head
        q.head = q.head.next
        del temp
        return x

if __name__ == '__main__':
    q = Queue()

    while(1):
        choice = input("\nWhat do you want to do?\n1) Insert integer.\n2) Pop integer.\n3) Exit\n")
        if(choice == "1"):
            insertInt = input("Enter integer to insert: ")
            a = Node(insertInt)
            enqueue(q, a)
            print("Inserted. Queue is now: ")
            traversal(q)
        elif(choice == "2"):
            print("Popping first in queue...")
            dequeue(q)
            print("Popped. Queue is now:")
            traversal(q)
        else:
            exit(0)
